Add example of using a UDF to predict values based on a model (#45)
arrowrbook/book: advanced_topics.qmd

{{< include _knitr_hooks.qmd >}}

```{r}
#| label: setup-advanced-data-manip
#| include: FALSE
library(arrow)
library(dplyr)
library(vctrs)
library(ggplot2)

> library(see)

Reviewer: What is this used for?

Reviewer 2: This is for the half violin. We could also copy pasta from stack overflow (and there's at least one other implementation), but I didn't want to (re)invent anything here

Reviewer: Ah ok cool. Maybe we can see:: reference the function we use there, just so it's clear in the code example that it's not a vanilla ggplot2 function?

  collect()
```

Another example is where you've created a model based on a subset of your data and you want to use it to make predictions.
Let's say we've created a linear model, using age and employment status to predict commute time for respondents in California in 2005.

```{r}
#| label: predict-commute-time
model_data <- open_dataset("data/person/") |>

>   filter(year==2005, location=="ca") |>

Reviewer: Same below too

```suggestion
  filter(year == 2005, location == "ca") |>
```

  transmute(
    commute_time = JWMNP,
    age = AGEP,
    self_employed = stringr::str_detect(as.character(COW), "Self-employed")
  ) |>
  collect()

my_model <- lm(commute_time ~ age + self_employed, data = model_data)
summary(my_model)
```

Let's say we want to use this model to predict commute times in California in 2021.
We can create a UDF which calls `predict()` to add the predicted commute time value to the data.

```{r}
#| label: predict-commute-time
register_scalar_function(
  "predict_commute_time",
  function(context, age, self_employed) {
    predict(my_model, newdata = data.frame(age, self_employed))
  },
  in_type = schema(age = int32(), self_employed = boolean()),
  out_type = float64(),
  auto_convert = TRUE
)

open_dataset("data/person/") |>
  filter(year==2021, location=="ca") |>
  transmute(
    age = AGEP,
    self_employed = stringr::str_detect(as.character(COW), "Self-employed"),
    commute_time_predicted = predict_commute_time(age, self_employed)
  ) |>
  collect() |>
  head()
```

> 

Reviewer: Should we add some discussion? Gratuitous plot of predicted/fitted vs. actual values?

Author: I added it but the plot looks a bit shit as the model is shit :joy:
Not sure if it's worth putting in the effort to make the model better to produce a better plot.
Let me know what you think!

Reviewer: I wonder if it would be worth 15 minutes of effort to try to make a reasonable model. Suppose that: calling back to the example in the Getting Started chapter, we fit a model where the outcome is whether someone works from home. We fit the model with 2015 data or something like we do here, and then show that it reasonably well predicts the outcome in 2019. Then we show with 2020 and how it doesn't. 

I don't think this is worth hours of effort, but it would be a nice story if it weren't hard to do. (And think of the page count from adding 3 new plots!)

Thoughts?

Author: It'll take _me_ more than 15 minutes as I'll end up agonising over exactly how good the model needs to be, lol, but if you take this one and I take the "produce the PDF in greyscale" instead, that works for me?

### How do UDFs work under the hood?

UDFs only work on scalar functions---functions which take one row of data as input and return one row of data as output.
Acero does not currently support user-defined aggregation functions.
For example, you could create a UDF which calculates the square of a number, but you can't create one which calculates the median value in a column of data.
These functions also cannot depend on other values, so you can't write a UDF which implements `lag()` or `lead()` which calculate a value based on the previous or next row in the data.

Because the UDF code is run in your R session, you won't be able to take advantage of some of the features of arrow, such as being able to run code on multiple CPU cores, or the faster speed of executing C++ code.

> You'll also be slowed down a little by the fact that arrow has to convert back and forth between R and arrow, and R is single-threaded, so each calculation must be run one at a time instead of allowing parallelization.

Reviewer: ```suggestion
You'll also be slowed down a little by the fact that arrow has to convert back and forth between R and arrow, and R is single-threaded, so each calculation is run on one chunk of data at a time instead of parallelized across many chunks at once.
```

  collect()
```

### Using UDFs for regression model predictions

Another example where UDFs can be useful is where you've created a model based on a subset of your data and you want to use it to make predictions.
Let's say we've created a linear model, using education and employment status to predict hours worked per week for respondents in 2018.

```{r}
#| label: model-commute-time
model_data <- open_dataset("data/person/") |>
    filter(year == 2018) |>
    filter(AGEP >= 16, AGEP < 65, !is.na(COW)) |>
    transmute(
      self_employed = str_detect(as.character(COW), "Self-employed"),
      edu = case_when(
        grepl("bachelor|associate|professional|doctorate|master", SCHL, TRUE) ~
          "College or higher",
        grepl("college", SCHL, TRUE) ~
          "Some college",
        grepl("High school graduate|high school diploma|GED", SCHL, TRUE) ~
          "High school",
        TRUE ~ "Less than High school"
      ),
      hours_worked = WKHP,
      PWGTP = PWGTP
    ) |>
    collect()

> hours_worked_model <- survey::svyglm(

Reviewer: We should explain using the `survey` package I think--I assume that's for the weighted regression?

Reviewer 2: Ack yes definitely. I had on my mental todo list to do that but didn't get to it. It is indeed to do the weighted regression

Reviewer 2: Done (not yet pushed)

  hours_worked ~ self_employed * edu,
  design = survey::svydesign(ids = ~1, weights = ~PWGTP, data = model_data)
)
summary(hours_worked_model)
```

We can use this model to predict what our model would have on this data.
We use the `predict` function to make these predictions and we also calculate the weighted mean for each group.
This is going to be very very close to the actual weighted means since this data was used to fit this linear model.

```{r}
#| label: summarise_fit_data
#| warning: false
model_data$predicted <- predict(hours_worked_model, newdata = model_data)

model_data_summary <- model_data |>
  group_by(edu, self_employed) |>
  summarise(
    work_hours_mean = weighted.mean(hours_worked, w = PWGTP, na.rm = TRUE),
    work_hours_predicted = weighted.mean(predicted, w = PWGTP),
  )
```

> And we can visualize the underlying data along with weighted mean and predictions.

Reviewer: This is me being 👴 I think but I hate this use of the word "visualize". Visualizing to me means closing your eyes and imagining something.

Reviewer 2: Heh, happy for any other verb

plot
```
:::

We are slightly off in most categories, but especially far off for high school and some college educated workers where we predict much lower working hours.
We can also re-use this UDF for any subset, including predicting on all of the US in 2021.

```{r}
#| label: predict-commute-time-all-us
modelled_data <- open_dataset("data/person/") |>
  filter(year == 2021) |>
  filter(AGEP >= 16, AGEP < 65, !is.na(COW)) |>
    transmute(
      self_employed = str_detect(as.character(COW), "Self-employed"),
      edu = case_when(
        grepl("bachelor|associate|professional|doctorate|master", SCHL, TRUE) ~
          "College or higher",
        grepl("college", SCHL, TRUE) ~
          "Some college",
        grepl("High school graduate|high school diploma|GED", SCHL, TRUE) ~
          "High school",
        TRUE ~ "Less than High school"
      ),
      hours_worked = WKHP,
      PWGTP = PWGTP,
      predicted = predict_work_hours(edu, self_employed),
    ) |>
  collect()

modelled_data |> head()
```

> And how good were our predictions?

Reviewer: After the plot, we should add a paragraph that answers this question.

Reviewer 2: Done. This actually _was here_ it was just in the wrong place (up about Alaska) 🤦 

I've also added a bit more narrative to these chunks

  collect()
```

### Using UDFs for regression model predictions

Another example where UDFs can be useful is where you've created a model based on a subset of your data and you want to use it to make predictions.
Let's say we've created a linear model, using education and employment status to predict hours worked per week for respondents in 2018.
When creating this model we aren't using the standard `stats::lm()` because our data is weighted. 

> the `survey` package has an equivalent regression function `survey::svyglm()` which applies the weights in the weight column during the regression. 

Author: ```suggestion
The `survey` package has an equivalent regression function `survey::svyglm()` which applies the weights in the weight column during the regression. 
```

This isn't strictly necesary for the demonstration here, but it does show how UDFs work not just with base R functions or functions in the tidyverse, but any R code that you come up with.

```{r}
#| label: model-commute-time
model_data <- open_dataset("data/person/") |>
    filter(year == 2018) |>
    filter(AGEP >= 16, AGEP < 65, !is.na(COW)) |>
    transmute(
      self_employed = str_detect(as.character(COW), "Self-employed"),
      edu = case_when(
        grepl("bachelor|associate|professional|doctorate|master", SCHL, TRUE) ~
          "College or higher",
        grepl("college", SCHL, TRUE) ~
          "Some college",
        grepl("High school graduate|high school diploma|GED", SCHL, TRUE) ~
          "High school",
        TRUE ~ "Less than High school"
      ),
      hours_worked = WKHP,
      PWGTP = PWGTP
    ) |>
    collect()

hours_worked_model <- survey::svyglm(
  hours_worked ~ self_employed * edu,
  design = survey::svydesign(ids = ~1, weights = ~PWGTP, data = model_data)
)
summary(hours_worked_model)
```

> We can use this model to predict what our model would have on this data.

Author: This sentence doesn't quite scan for me, but wasn't sure what to change it to sorry

Reviewer 2: Aaaah that's because it's half complete. I added the bit I had in my head but didn't actually write out. I also rearranged and expanded a little bit more detail around there too.

We use the `predict` function to make these predictions and we also calculate the weighted mean for each group.
This is going to be very close to the actual weighted means since this data was used to fit this linear model.

```{r}
#| label: summarise_fit_data
#| warning: false
model_data$predicted <- predict(hours_worked_model, newdata = model_data)

model_data_summary <- model_data |>
  group_by(edu, self_employed) |>
  summarise(
    work_hours_mean = weighted.mean(hours_worked, w = PWGTP, na.rm = TRUE),
    work_hours_predicted = weighted.mean(predicted, w = PWGTP),
  )
```

And we can plot our model as well as the predictions to visualize what is going on.

> Here the underlying data is represented in teh density curve --- the taller the curve the more observations are at that point on the x-axis (hours worked per week). 

Author: ```suggestion
Here the underlying data is represented in the density curve --- the taller the curve the more observations are at that point on the x-axis (hours worked per week). 
```

  ) +
  layer(
    geom = "point",
    stat = "identity",
    position = position_nudge(x = -0.08),
    data = modelled_data_summary,
    mapping = aes(x = edu, y = work_hours_predicted, weight = NULL),
    params = list(shape = 18, size = 4)
  ) +
  facet_grid(
    cols = vars(self_employed),
    labeller = as_labeller(c("TRUE" = "self-employed", "FALSE" = "not self-employed"))
  ) +
  coord_flip() +
  theme_minimal() +
    labs(
      title = "Hours worked per week",
      subtitle = "by employment type in 2018, all of US\nvertical line is weighted mean, diamond model prediciton",
      x = "education",
      y = "hours worked per week"
    )
```

::: {.content-visible when-format="html"}
```{r}
#| label: plot_predictions_all_us_plot
#| echo: false
#| warning: false
plot
```
:::

::: {.content-visible when-format="pdf"}
```{r}
#| label: plot_predictions_all_us_plot-pdf
#| echo: false
#| warning: false
#| fig-height: 9
#| fig-cap: "Hours worked per week by employment type in 2021 for all of the United States. The density curve is the raw data, the line at the bottom the weighted mean and the diamond at the bottom is the weighted mean of the model predictions."
plot
```
:::

> And we can see the fit is slightly better, though we still slightly over predict particularly self-employed workers.

Author: ```suggestion
And we can see the fit is slightly better, though we still slightly over predict, particularly for self-employed workers.
```

plot <- ggplot(model_data) +
  aes(
    x = factor(

>       edu, 

Reviewer 2: Ooph apparently I don't (yet) have positron on my desktop setup for doing this on save :/ I should do that

::: {.content-visible when-format="html"}
```{r}
#| label: plot_fit_data_plot
#| warning: false
#| echo: false
plot
```
:::

::: {.content-visible when-format="pdf"}
```{r}
#| label: plot_fit_data_plot-pdf
#| echo: false

> #| warning: false

Reviewer 2: Thanks for adding this. I realized when I looked at the preview I had it but then when I split web vs pdf I copy pastaed it out of existence and hadn't remembered to put it back
